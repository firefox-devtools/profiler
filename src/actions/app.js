/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { oneLine } from 'common-tags';
import {
  getSelectedTab,
  getDataSource,
  getLocalTrackOrderByPid,
} from 'firefox-profiler/selectors/url-state';
import {
  getTrackThreadHeights,
  getIsEventDelayTracksEnabled,
  getIsExperimentalCPUGraphsEnabled,
  getIsExperimentalProcessCPUTracksEnabled,
} from 'firefox-profiler/selectors/app';
import {
  getLocalTracksByPid,
  getThreads,
  getCounters,
} from 'firefox-profiler/selectors/profile';
import { sendAnalytics } from 'firefox-profiler/utils/analytics';
import {
  stateFromLocation,
  withHistoryReplaceStateSync,
} from 'firefox-profiler/app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { fatalError } from './errors';
import {
  addEventDelayTracksForThreads,
  initializeLocalTrackOrderByPid,
  addProcessCPUTracksForProcess,
} from 'firefox-profiler/profile-logic/tracks';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getIsCPUUtilizationProvided,
  getAreThereAnyProcessCPUCounters,
} from 'firefox-profiler/selectors/cpu';

import type {
  Profile,
  ThreadsKey,
  CssPixels,
  Action,
  ThunkAction,
  UrlState,
  UploadedProfileInformation,
  IndexIntoCategoryList,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnection,
  BrowserConnectionStatus,
} from 'firefox-profiler/app-logic/browser-connection';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 * `profile` parameter can be null when the data source can't provide the profile
 * and the url upgrader step is not needed (e.g. 'from-browser').
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile | null,
  browserConnection: BrowserConnection | null
): ThunkAction<void> {
  return (dispatch) => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      if (e.name === 'UrlUpgradeError') {
        // The error is an URL upgrade error, let's fire a fatal error.
        // If there's a service worker update, the class `ServiceWorkerManager`
        // will automatically reload in case the new code knows how to handle
        // this URL version.
        dispatch(fatalError(e));
        return;
      }
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    // All of this is done while the history is replaced, as this is part of the initial
    // load process.
    withHistoryReplaceStateSync(() => {
      dispatch(updateUrlState(urlState));
      dispatch(finalizeProfileView(browserConnection));
      dispatch(urlSetupDone());
    });
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadsKey: ThreadsKey,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadsKey];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadsKey,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}

/*
 * This action enables the event delay tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableEventDelayTracks()`
 */
export function enableEventDelayTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsEventDelayTracksEnabled(getState())) {
      console.error(
        'Tried to enable the event delay tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      selectedThreadSelectors.getSamplesTable(getState()).eventDelay ===
      undefined
    ) {
      // Return early if the profile doesn't have eventDelay values.
      console.error(oneLine`
        Tried to enable the event delay tracks, but this profile does
        not have eventDelay values. It is likely an older profile.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addEventDelayTracksForThreads(
      getThreads(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );
    dispatch({
      type: 'ENABLE_EVENT_DELAY_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/*
 * This action enables the CPU graph tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableCPUGraphs()`
 */
export function enableExperimentalCPUGraphs(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalCPUGraphsEnabled(getState())) {
      console.error(
        'Tried to enable the CPU graph tracks, but they are already enabled.'
      );
      return false;
    }

    if (!getIsCPUUtilizationProvided(getState())) {
      // Return early if the profile doesn't have threadCPUDelta values.
      console.error(oneLine`
        Tried to enable the CPU graph tracks, but this profile does
        not have threadCPUDelta values. It is likely an older profile.
      `);
      return false;
    }

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_CPU_GRAPHS',
    });

    return true;
  };
}

/*
 * This action enables the process CPU tracks. They are hidden by default because
 * the front-end work is not done for this data yet.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableExperimentalProcessCPUTracks()`
 */
export function enableExperimentalProcessCPUTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalProcessCPUTracksEnabled(getState())) {
      console.error(
        'Tried to enable the process CPU tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      !getIsCPUUtilizationProvided(getState()) &&
      getAreThereAnyProcessCPUCounters(getState())
    ) {
      // Return early if the profile doesn't have threadCPUDelta values or
      // doesn't have any experimental process CPU counters.
      console.error(oneLine`
        Tried to enable the process CPU tracks, but this profile does
        not have threadCPUDelta values or process CPU threads.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addProcessCPUTracksForProcess(
      getCounters(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_PROCESS_CPU_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/**
 * This caches the profile data in the local state for synchronous access.
 */
export function setCurrentProfileUploadedInformation(
  uploadedProfileInformation: UploadedProfileInformation | null
): Action {
  return {
    type: 'SET_CURRENT_PROFILE_UPLOADED_INFORMATION',
    uploadedProfileInformation,
  };
}

export function profileRemotelyDeleted(): Action {
  // Ideally we should store the current profile data in a local indexeddb, and
  // set the URL to /local/<indexeddb-key>.
  return { type: 'PROFILE_REMOTELY_DELETED' };
}

export function updateBrowserConnectionStatus(
  browserConnectionStatus: BrowserConnectionStatus
): Action {
  return { type: 'UPDATE_BROWSER_CONNECTION_STATUS', browserConnectionStatus };
}

export function toggleOpenCategoryInSidebar(
  kind: string,
  category: IndexIntoCategoryList
): Action {
  return {
    type: 'TOGGLE_SIDEBAR_OPEN_CATEGORY',
    kind,
    category,
  };
}
