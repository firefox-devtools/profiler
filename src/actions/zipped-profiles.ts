/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import {
  getZipFileTable,
  getZipFileState,
} from 'firefox-profiler/selectors/zipped-profiles';
import { unserializeProfileOfArbitraryFormat } from 'firefox-profiler/profile-logic/process-profile';
import { loadProfile } from './receive-profile';

import { Action, ThunkAction } from 'firefox-profiler/types';
import { IndexIntoZipFileTable } from 'firefox-profiler/profile-logic/zip-files';

export function changeSelectedZipFile(
  selectedZipFileIndex: IndexIntoZipFileTable
): Action {
  return {
    type: 'CHANGE_SELECTED_ZIP_FILE' as const,
    selectedZipFileIndex,
  };
}

export function changeExpandedZipFile(
  expandedZipFileIndexes: Array<IndexIntoZipFileTable | null>
): Action {
  return {
    type: 'CHANGE_EXPANDED_ZIP_FILES' as const,
    expandedZipFileIndexes,
  };
}

/**
 * This ThunkAction has a bit of complexity to it, due to the asynchronous nature of
 * reading in profiles from a zip file. The UrlState represents the current desired state
 * of what file to view in a zip file, and the ZipFileState represents the actual steps
 * being performed to reach the desired UrlState.
 *
 * This ThunkAction needs to properly handle when the UrlState changes faster than it
 * can change the ZipFileState, and not have any race conditions.
 */
export function viewProfileFromZip(
  zipFileIndex: IndexIntoZipFileTable,
  initialLoad: boolean = false
): ThunkAction<Promise<void>> {
  return async (dispatch, getState) => {
    const zipFileTable = getZipFileTable(getState());
    const pathInZipFile = zipFileTable.path[zipFileIndex];
    const file = zipFileTable.file[zipFileIndex];
    if (!file) {
      throw new Error(
        'Attempted to load an archive that did not exist or was a directory.'
      );
    }

    dispatch({ type: 'PROCESS_PROFILE_FROM_ZIP_FILE' as const, pathInZipFile });

    try {
      // Attempt to unserialize the profile.
      const profile = await unserializeProfileOfArbitraryFormat(
        await file.async('string'),
        pathInZipFile
      );

      // Since this is an async function, there can be race conditions. Prevent this by
      // comparing this request with the current state of the store. If this result
      // is invalid, don't dispatch anything, and discard the profile.
      const zipFileState = getZipFileState(getState());
      if (
        zipFileState.pathInZipFile === pathInZipFile &&
        zipFileState.phase === 'PROCESS_PROFILE_FROM_ZIP_FILE'
      ) {
        await dispatch(loadProfile(profile, { pathInZipFile }, initialLoad));
      }
    } catch (error) {
      console.error(
        'Failed to process the profile in the archive with the following error:'
      );
      console.error(error);
      dispatch({
        type: 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE' as const,
        error,
      });
    }
  };
}

/**
 * This function can take a zip file path, but the path can come from the URL, so
 * don't really trust it.
 */
export function viewProfileFromPathInZipFile(
  pathInZipFile: string
): ThunkAction<Promise<void>> {
  return (dispatch, getState) => {
    const zipFileTable = getZipFileTable(getState());
    const zipFileIndex = zipFileTable.path.indexOf(pathInZipFile);
    if (zipFileIndex === -1) {
      dispatch(showErrorForNoFileInZip(pathInZipFile));
      return Promise.resolve();
    }
    return dispatch(viewProfileFromZip(zipFileIndex));
  };
}

export function returnToZipFileList() {
  return { type: 'RETURN_TO_ZIP_FILE_LIST' as const };
}

export function showErrorForNoFileInZip(pathInZipFile: string) {
  return { type: 'FILE_NOT_FOUND_IN_ZIP_FILE' as const, pathInZipFile };
}
