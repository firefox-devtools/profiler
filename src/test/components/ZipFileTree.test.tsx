/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { ZipFileViewer } from '../../components/app/ZipFileViewer';
import { Provider } from 'react-redux';

import {
  render,
  screen,
  act,
} from 'firefox-profiler/test/fixtures/testing-library';
import * as UrlStateSelectors from '../../selectors/url-state';
import * as ZippedProfileSelectors from '../../selectors/zipped-profiles';

import { storeWithZipFile } from '../fixtures/profiles/zip-file';
import { autoMockCanvasContext } from '../fixtures/mocks/canvas-context';
import { autoMockElementSize } from '../fixtures/mocks/element-size';
import { waitUntilState, fireFullClick } from '../fixtures/utils';
import { autoMockIntersectionObserver } from '../fixtures/mocks/intersection-observer';

describe('calltree/ZipFileTree', function () {
  autoMockCanvasContext();
  autoMockIntersectionObserver();

  // This makes the bounding box large enough so that we don't trigger
  // VirtualList's virtualization.
  autoMockElementSize({ width: 1000, height: 2000 });

  async function setup() {
    const { store } = await storeWithZipFile([
      'foo/bar/profile1.json',
      'foo/profile2.json',
      'baz/profile3.json',
      // Use a file with a big depth to test the automatic expansion at load time.
      'boat/ship/new/anything/explore/yes/profile4.json',
      'not/a/profile.pdf',
    ]);

    const renderResult = render(
      <Provider store={store}>
        <ZipFileViewer />
      </Provider>
    );
    const { getState, dispatch } = store;
    return { ...renderResult, store, getState, dispatch };
  }

  it('renders a zip file tree', async () => {
    const { container } = await setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('contains a list of all the files', async () => {
    await setup();

    // We're looping through all expected files and check if we can find an
    // element with the file name as text content.
    [
      'foo',
      'bar',
      'profile1',
      'profile2',
      'baz',
      'profile3',
      'profile4',
    ].forEach((fileName) =>
      expect(screen.getByText(fileName)).toBeInTheDocument()
    );
  });

  it('removes .json extensions', async () => {
    await setup();

    ['profile1', 'profile2', 'profile3', 'profile4'].forEach((fileName) =>
      expect(screen.getByText(fileName)).toBeInTheDocument()
    );

    [
      'profile1.json',
      'profile2.json',
      'profile3.json',
      'profile4.json',
    ].forEach((fileName) =>
      expect(screen.queryByText(fileName)).not.toBeInTheDocument()
    );
  });

  it('preserves extensions other than .json', async () => {
    await setup();

    expect(screen.getByText('profile.pdf')).toBeInTheDocument();
  });

  describe('clicking on a profile link', function () {
    const setupClickingTest = async () => {
      const setupResult = await setup();
      const { store, container } = setupResult;

      const profile1OpenLink = screen.getByText('profile1');

      const waitUntilDoneProcessingZip = () =>
        act(() =>
          waitUntilState(
            store,
            (state) =>
              ZippedProfileSelectors.getZipFileState(state).phase ===
              'VIEW_PROFILE_IN_ZIP_FILE'
          )
        );

      function profileViewer() {
        return container.querySelector('.profileViewer');
      }

      return {
        ...setupResult,
        profile1OpenLink,
        waitUntilDoneProcessingZip,
        profileViewer,
      };
    };

    it('starts out without any path in the zip file', async () => {
      const { getState, profileViewer } = await setupClickingTest();
      expect(UrlStateSelectors.getPathInZipFileFromUrl(getState())).toBe(null);
      expect(profileViewer()).toBeFalsy();
    });

    it('kicks off the profile loading process when clicked', async () => {
      const {
        getState,
        profile1OpenLink,
        waitUntilDoneProcessingZip,
        profileViewer,
      } = await setupClickingTest();
      fireFullClick(profile1OpenLink);
      expect(UrlStateSelectors.getPathInZipFileFromUrl(getState())).toBe(
        'foo/bar/profile1.json'
      );

      // The profile isn't actually viewed yet.
      expect(profileViewer()).toBeFalsy();
      expect(
        screen.getByText(/Loading the profile/).textContent
      ).toMatchSnapshot();

      // There is async behavior going on. Wait until it is done or else
      // redux-react will throw an error.
      await waitUntilDoneProcessingZip();
    });

    it('will show the ProfileViewer component when done processing', async () => {
      const { profileViewer, profile1OpenLink, waitUntilDoneProcessingZip } =
        await setupClickingTest();

      fireFullClick(profile1OpenLink);
      await waitUntilDoneProcessingZip();
      expect(profileViewer()).toBeTruthy();
    });
  });
});
