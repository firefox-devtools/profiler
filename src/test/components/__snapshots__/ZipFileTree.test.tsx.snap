// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`calltree/ZipFileTree clicking on a profile link kicks off the profile loading process when clicked 1`] = `"Loading the profile from the archive..."`;

exports[`calltree/ZipFileTree renders a zip file tree 1`] = `
<section
  class="zipFileViewer"
>
  <div
    class="zipFileViewerSection"
  >
    <header
      class="zipFileViewerHeader"
    >
      <h1>
        Firefox Profiler
      </h1>
      <p>
        Choose a profile from this archive
      </p>
    </header>
    <div
      class="treeView"
    >
      <div
        class="react-contextmenu-wrapper treeViewContextMenu"
      >
        <div
          aria-label="Call tree"
          class="treeViewBody"
          role="tree"
          tabindex="0"
        >
          <div
            class="treeViewBodyInnerWrapper"
          >
            <div
              class="treeViewBodyInner treeViewBodyInner0"
              style="height: 510px;"
            >
              <div
                class="treeViewBodyInner treeViewBodyInner0TopSpacer"
                style="height: 0px;"
              />
              <div
                class="treeViewBodyInner treeViewBodyInner0InnerChunk"
              >
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even"
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd"
                  style="height: 30px; line-height: 30px;"
                />
              </div>
            </div>
            <div
              class="treeViewBodyInner treeViewBodyInner1"
              style="height: 510px; min-width: 3000px;"
            >
              <div
                class="treeViewBodyInner treeViewBodyInner1TopSpacer"
                style="height: 0px;"
              />
              <div
                class="treeViewBodyInner treeViewBodyInner1InnerChunk"
              >
                <div
                  aria-expanded="true"
                  aria-level="1"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-0"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 0px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    foo
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-1"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    bar
                  </span>
                </div>
                <div
                  aria-level="3"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-2"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 30px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=foo%2Fbar%2Fprofile1.json"
                    >
                      profile1
                    </a>
                  </span>
                </div>
                <div
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-3"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=foo%2Fprofile2.json"
                    >
                      profile2
                    </a>
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="1"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-4"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 0px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    baz
                  </span>
                </div>
                <div
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-5"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=baz%2Fprofile3.json"
                    >
                      profile3
                    </a>
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="1"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-6"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 0px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    boat
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-7"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    ship
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="3"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-8"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 30px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    new
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="4"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-9"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 45px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    anything
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="5"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-10"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 60px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    explore
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="6"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-11"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 75px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    yes
                  </span>
                </div>
                <div
                  aria-level="7"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-12"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 90px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=boat%2Fship%2Fnew%2Fanything%2Fexplore%2Fyes%2Fprofile4.json"
                    >
                      profile4
                    </a>
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="1"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-13"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 0px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    not
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even"
                  id="treeViewRow-14"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    a
                  </span>
                </div>
                <div
                  aria-level="3"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd"
                  id="treeViewRow-15"
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 30px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=not%2Fa%2Fprofile.pdf"
                    >
                      profile.pdf
                    </a>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="dragAndDropOverlayWrapper"
    >
      <div
        class="dragAndDropOverlay"
      >
        Drop a saved profile here
      </div>
    </div>
  </div>
</section>
`;
